In [ ]:
import pandas as pd

pd.set_option("display.max_colwidth", 150)

caminho = "../data/raw/"

# lendo os datasets, vamos focar no reviews
# aproveitando para ler todos os datasets caso sobre tempo para o desafio
order = pd.read_csv(caminho + "olist_orders_dataset.csv")
reviews = pd.read_csv(caminho + "olist_order_reviews_dataset.csv")
items = pd.read_csv(caminho + "olist_order_items_dataset.csv")
products = pd.read_csv(caminho + "olist_products_dataset.csv")
customers = pd.read_csv(caminho + "olist_customers_dataset.csv")

print("orders shape: ", order.shape)
print("reviews shape: ", reviews.shape)
print("items shape: ", items.shape)
print("products shape: ", products.shape)
print("customers shape: ", customers.shape)

orders shape:  (99441, 8)
reviews shape:  (99224, 7)
items shape:  (112650, 7)
products shape:  (32951, 9)
customers shape:  (99441, 5)


In [2]:
# verificando dataset de reviews
reviews.info()
reviews.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-null  object
 4   review_comment_message   40977 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB


,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela Internet seguro e prático Parabéns a todos feliz Páscoa,2018-03-01 00:00:00,2018-03-02 10:26:53
5,15197aa66ff4d0650b5434f1b46cda19,b18dcdf73be66366873cd26c5724d1dc,1,NaN,NaN,2018-04-13 00:00:00,2018-04-16 00:39:37
6,07f9bee5d1b850860defd761afa7ff16,e48aa0d2dcec3a2e87348811bcfdf22b,5,NaN,NaN,2017-07-16 00:00:00,2017-07-18 19:30:34
7,7c6400515c67679fbee952a7525281ef,c31a859e34e3adac22f376954e19b39d,5,NaN,NaN,2018-08-14 00:00:00,2018-08-14 21:36:06
8,a3f6f7f6f433de0aefbb97da197c554c,9c214ac970e84273583ab523dfafd09b,5,NaN,NaN,2017-05-17 00:00:00,2017-05-18 12:05:37
9,8670d52e15e00043ae7de4c01cc2fe06,b9bf720beb4ab3728760088589c62129,4,recomendo,aparelho eficiente. no site a marca do aparelho esta impresso como 3desinfector e ao chegar esta com outro nome...atualizar com a marca correta um...,2018-05-22 00:00:00,2018-05-23 16:45:47


In [ ]:
# analisando dataset
# total de nulos na coluna de review_comment_title e review_comment_message
total_avaliacoes = len(reviews)

print("Total de avaliações: ", total_avaliacoes)

total_nulos_title =reviews["review_comment_title"].isnull().sum()
print("Total de avaliações sem título: ", total_nulos_title)

total_nulos_message =reviews["review_comment_message"].isnull().sum()
print("Total de avaliações sem mensagem: ", total_nulos_message)

print("% mensagem nula:", round(reviews["review_comment_message"].isnull().mean() * 100, 1), "%")

Total de avaliações:  99224
Total de avaliações sem título:  87656
Total de avaliações sem mensagem:  58247
% mensagem nula: 58.7 %


In [ ]:
# analisando a distribuição das notas

# dataset somente com os reviews que possuem texto
com_texto = reviews[reviews["review_comment_message"].notnull()].copy()

print("Total de avaliações com texto: ", len(com_texto))
print("Distribuição de todas as notas")
print(reviews['review_score'].value_counts(normalize=True).sort_index() * 100)
print()
print("Distribuição de notas de quem comentou")
print(com_texto['review_score'].value_counts(normalize=True).sort_index() * 100)

Total de avaliações com texto:  40977
Distribuição de todas as notas
review_score
1    11.513344
2     3.175643
3     8.242965
4    19.291704
5    57.776344
Name: proportion, dtype: float64

Distribuição de notas de quem comentou
review_score
1    21.341240
2     5.234644
3     8.680479
4    14.583791
5    50.159846
Name: proportion, dtype: float64


as distribuicoes dos comentario por nota, tanto do dataset inteiro como do que tem somente comentarios preenchidos, parecem ter a mesma tendencia, observamos que as pessoas que mais comentam atribuem notas altas como 4 e 5, mas tambem as que atribuem notas ruins como a nota 1.

In [7]:
# analisando tamanho dos comentarios
com_texto["tamanho_caracteres"] = com_texto["review_comment_message"].str.len()
com_texto["tamanho_palavras"] = com_texto["review_comment_message"].str.split().str.len()

print(com_texto[["tamanho_caracteres", "tamanho_palavras"]].describe())

       tamanho_caracteres  tamanho_palavras
count        40977.000000      40977.000000
mean            68.637675         11.679210
std             53.849207          9.554347
min              1.000000          0.000000
25%             28.000000          4.000000
50%             53.000000          9.000000
75%             95.000000         16.000000
max            208.000000         45.000000


 As medias sao aproximadamente 69 caracteres e 12 palavras com no maximo 208 caracteres e 45 palavras por comentario.

 Parecem ser comentarios breves, 75% ficam com 16 palavras.

In [12]:
# vendo exemplos
print("Exemplo curto")
print(com_texto.sort_values("tamanho_palavras").iloc[0]["review_comment_message"])
print()
print(" Exemplo mediano")
meio = len(com_texto) // 2
print(com_texto.sort_values("tamanho_palavras").iloc[meio]["review_comment_message"])
print()
print(" Exemplo longo")
print(com_texto.sort_values("tamanho_palavras", ascending=False).iloc[0]["review_comment_message"])

Exemplo curto
 

 Exemplo mediano
Cortina como eu esperava, acabamento bonito. Atendei minhas expectativas. 

 Exemplo longo
Eu pedi 110 e veio 220 e não consigo fazer a troca ja liguei mil vezes ninguém me atende o site é uma merda não tem opção de troca... E eu fiz uma compra de uns bancos carros de boi foi um golpe q cai


In [20]:
com_texto.sort_values("tamanho_palavras", ascending=True)

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp,tamanho_mensagem,tamanho_palavras,tamanho_caracteres
75383,8e2d665afbaf7f089558d04f6d4ce7f9,774906e7da9df9ff1e2a2aa1dda5034b,5,NaN,,2018-04-17 00:00:00,2018-04-17 18:45:48,1,0,1
30049,5911592a4b0c81d4936fab9aedcd6426,a6cd7cac7ee11538a78649066487e0ca,4,Recomendo,\r\n,2018-07-27 00:00:00,2018-07-30 20:49:41,2,0,2
45138,04bb4510dfa4af74e357521612d82ae9,40ccb6520c6922d1712766b66922e72d,5,NaN,,2018-04-11 00:00:00,2018-04-12 02:28:17,1,0,1
92751,b2c19dad0f5ff7f2cd1c4a1d27646820,71c3e399882c5bab8e8b576ee1558da3,5,NaN,\r\n,2017-12-13 00:00:00,2017-12-13 22:19:33,2,0,2
26041,b905a83cd40f67b08363261fc2d2dadb,cf65e1cf310d6860d96e2efaab1f5d75,5,NaN,\r\n,2017-08-22 00:00:00,2017-08-25 01:08:15,2,0,2
...,...,...,...,...,...,...,...,...,...,...
40984,ea98b47899516adbcf3bfe4f7e491587,cf650037c918409bca8e5e67ee7dcfbe,1,Relógio ruim,Propaganda enganosa lá na descrição diz que o relógio é isso e aquilo que é bom e tal mas na verdade não é nada disso além de tudo a cor que eu pe...,2018-07-04 00:00:00,2018-07-04 19:47:19,194,43,194
73913,297292ce8b330c2b8f7e330c7dd06079,446d3eee7dfa0cb1844584b53bc37cab,5,NaN,Gostaria de agradecer pois o produto veio em perfeito estado e sem contar com o prazo de entrega. Efetuei a compra no dia 31 de Maio o prazo de en...,2017-06-07 00:00:00,2017-06-08 17:24:45,200,43,200
67175,9aa0f0428e7491e8bcf0f21626119b88,665529c259350d60618eca48496ea75d,1,NaN,Eu pedi 110 e veio 220 e não consigo fazer a troca ja liguei mil vezes ninguém me atende o site é uma merda não tem opção de troca... E eu fiz uma...,2017-12-22 00:00:00,2017-12-22 22:50:34,200,45,200
26343,c9bad86ea532ed43c5a21d38f90f3255,3ed99ad145f818a65b1bcfce6a4182fb,3,NaN,"Produto bem mais ou menos . Prazo : OK entrega: OK . porém , o produto parece de camelô , não é bonita , nem parece que comprei em loja . o matéri...",2018-01-30 00:00:00,2018-01-30 19:19:01,197,45,197


Ao analisar mais a fundo e visualizar alguns exemplos, notamos que existem strings vazias que n estao marcadas como nulas, por isso o describe de quantidade de palavras retornou 0 como menor comentario.

In [22]:
# quantidade de comentarios quando retiramos os nulos
com_texto.shape

(40977, 10)

In [23]:
# de fato quantos comentarios temos retirando tmb as strings vazias
com_texto["texto_stripped"] = com_texto["review_comment_message"].str.strip()

vazios_disfarcados = (com_texto["texto_stripped"] == "").sum()
print("Comentários 'string vazias':", vazios_disfarcados)
print("% sobre quem 'tinha texto':", round(vazios_disfarcados / len(com_texto) * 100, 2), "%")
print()
com_texto_real = com_texto[com_texto["texto_stripped"] != ""].copy()
print("Total de avaliações com texto REAL, após limpeza:", len(com_texto_real))

Comentários 'string vazias': 27
% sobre quem 'tinha texto': 0.07 %

Total de avaliações com texto REAL, após limpeza: 40950


In [24]:
com_texto_real["tamanho_palavras_real"] = com_texto_real["texto_stripped"].str.split().str.len()

print(com_texto_real["tamanho_palavras_real"].describe())
print()
print("Comentários com 1 palavra só:")
print(com_texto_real[com_texto_real["tamanho_palavras_real"] == 1]["texto_stripped"].head(10).tolist())

count    40950.000000
mean        11.686911
std          9.552786
min          1.000000
25%          4.000000
50%          9.000000
75%         16.000000
max         45.000000
Name: tamanho_palavras_real, dtype: float64

Comentários com 1 palavra só:
['Péssimo', 'boa', 'Maravilha', 'Satisfeito', 'Gostei', 'OTIMA', 'OK', 'Recomendo', 'otimo', 'A']


O volume de comentarios com string vazias e bem baixo de 0.07%, mesmo assim vale retirar tmb estes ruidos.

A base fica entao com 40950 registros.

Tambem verificamos os comentarios com apenas uma palavra, parecem contribuir para uma tendencia positiva ou negativa.

In [25]:
com_texto_real

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp,tamanho_mensagem,tamanho_palavras,tamanho_caracteres,texto_stripped,tamanho_palavras_real
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06,37,6,37,Recebi bem antes do prazo estipulado.,6
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela Internet seguro e prático Parabéns a todos feliz Páscoa,2018-03-01 00:00:00,2018-03-02 10:26:53,100,15,100,Parabéns lojas lannister adorei comprar pela Internet seguro e prático Parabéns a todos feliz Páscoa,15
9,8670d52e15e00043ae7de4c01cc2fe06,b9bf720beb4ab3728760088589c62129,4,recomendo,aparelho eficiente. no site a marca do aparelho esta impresso como 3desinfector e ao chegar esta com outro nome...atualizar com a marca correta um...,2018-05-22 00:00:00,2018-05-23 16:45:47,174,30,174,aparelho eficiente. no site a marca do aparelho esta impresso como 3desinfector e ao chegar esta com outro nome...atualizar com a marca correta um...,30
12,4b49719c8a200003f700d3d986ea1a19,9d6f15f95d01e79bd1349cc208361f09,4,NaN,"Mas um pouco ,travando...pelo valor ta Boa.\r\n",2018-02-16 00:00:00,2018-02-20 10:52:22,45,7,45,"Mas um pouco ,travando...pelo valor ta Boa.",7
15,3948b09f7c818e2d86c9a546758b2335,e51478e7e277a83743b6f9991dbfa3fb,5,Super recomendo,"Vendedor confiável, produto ok e entrega antes do prazo.",2018-05-23 00:00:00,2018-05-24 03:00:01,56,9,56,"Vendedor confiável, produto ok e entrega antes do prazo.",9
...,...,...,...,...,...,...,...,...,...,...,...,...
99205,98fffa80dc9acbde7388bef1600f3b15,d398e9c82363c12527f71801bf0e6100,4,NaN,para este produto recebi de acordo com a compra realizada!,2017-11-29 00:00:00,2017-11-30 15:52:51,58,10,58,para este produto recebi de acordo com a compra realizada!,10
99208,df5fae90e85354241d5d64a8955b2b09,509b86c65fe4e2ad5b96408cfef9755e,5,NaN,Entregou dentro do prazo. O produto chegou em condições perfeitas. Estou muito satisfeita!,2018-02-07 00:00:00,2018-02-19 19:47:23,91,13,91,Entregou dentro do prazo. O produto chegou em condições perfeitas. Estou muito satisfeita!,13
99215,a709d176f59bc3af77f4149c96bae357,d5cb12269711bd1eaf7eed8fd32a7c95,3,NaN,"O produto não foi enviado com NF, não existe venda sem NF, com certeza fico no aguardo do envio da NF podendo ser por e-mail.",2018-05-19 00:00:00,2018-05-20 21:51:06,125,25,125,"O produto não foi enviado com NF, não existe venda sem NF, com certeza fico no aguardo do envio da NF podendo ser por e-mail.",25
99221,b3de70c89b1510c4cd3d0649fd302472,55d4004744368f5571d1f590031933e4,5,NaN,"Excelente mochila, entrega super rápida. Super recomendo essa loja!",2018-03-22 00:00:00,2018-03-23 09:10:43,67,9,67,"Excelente mochila, entrega super rápida. Super recomendo essa loja!",9


In [ ]:
# analisando duplicidades
